Support/bli 618 separete json c from knot cloud

agms-cesar opened the pull request.

agms-cesar committed:
Removes hard-coded json fields from sdk

- Removes hard-coded json fields from knot_cloud.c and parser.c;
- Creates definitions for json fields in parser.h;
- Update parser methods to use json field definitions;
- Update knot_cloud to use json field definitions.

CaioCordeiro commented on src/knot_cloud.c:
```diff
@@ -516,15 +443,14 @@ int knot_cloud_auth_device(const char *id, const char *token)
 		return KNOT_ERR_CLOUD_FAILURE;
 
-	json_str = json_object_to_json_string(jobj_schema);
+	/*json_str = json_object_to_json_string(jobj_schema);*/
```
Is this comment really necessary?

CaioCordeiro commented on src/parser.c:
```diff
@@ -398,55 +478,63 @@ struct l_queue *parser_schema_to_list(const char *json_str)
 
 	for (i = 0; i < json_object_array_length(jobjarray); i++) {
-
+		has_err = true;
```
Please use tabs for indentation.

agms-cesar replied:
this is a tab i think

CaioCordeiro commented on src/parser.c:
```diff
@@ -494,53 +584,68 @@ struct l_queue *parser_queue_from_json_array(json_object *jobj,
 			l_queue_push_tail(list, item);
 	}
-
+	json_object_put(jobj);
```
Please use tabs for indentation.

agms-cesar replied:
I think this is a tab already

CaioCordeiro commented on src/parser.c:
```diff
@@ -494,53 +584,68 @@ struct l_queue *parser_queue_from_json_array(json_object *jobj,
 	json_object *json_array;
 	json_object *jobjentry;
-	int sensor_id;
+	int sensor_id;	
```
Please remove extra blank space.

CaioCordeiro commented on src/parser.c:
```diff
@@ -494,53 +584,68 @@ struct l_queue *parser_queue_from_json_array(json_object *jobj,
+		l_queue_destroy(list, l_free);
+		return NULL;
 	}
-
```
Is removing this line really necessary?

agms-cesar replied:
I think it is, because the next statement is a return

CaioCordeiro commented on src/parser.c:
```diff
@@ -582,74 +689,97 @@ json_object *parser_device_json_create(const char *device_id,
 
 	type = json_object_get_type(jobjkey);
-
+	json_object_put(jso);
```
Please use tabs for indentation.

agms-cesar replied:
I think this is already a tab.

CaioCordeiro reviewed (changes requested):
LGTM. I suggested some code formatting changes, but the code looks good.

CaioCordeiro commented on src/knot_cloud.c:
```diff
@@ -236,22 +182,11 @@ static bool on_amqp_receive_message(const char *exchange,
-
-	json_object_put(jso);
-
 	return consumed;
```
Its nice to have a blank line before a return statement.

CaioCordeiro commented on src/parser.c:
```diff
@@ -219,12 +269,14 @@ json_object *parser_schema_create_object(const char *device_id,
-	return json_msg;
+	json_str = l_strdup(json_object_to_json_string(json_msg));
+	json_object_put(json_msg);
+	return json_str;
```
Its nice to have a blank line before a return statement.

CaioCordeiro commented on src/parser.c:
```diff
@@ -550,29 +655,31 @@ json_object *parser_sensorid_to_json(const char *key, struct l_queue *list)
-json_object *parser_device_json_create(const char *device_id,
-				       const char *device_name)
+char *parser_device_json_create(const char *device_id,
+			        const char *device_name)
```
I think that a tab can fit in here.

CaioCordeiro commented on src/parser.c:
```diff
@@ -582,74 +689,101 @@ json_object *parser_device_json_create(const char *device_id,
-json_object *parser_auth_json_create(const char *device_id,
-				     const char *device_token)
+char *parser_auth_json_create(const char *device_id,
+			      const char *device_token)
```
I think that a tab can fit in here.

CaioCordeiro reviewed (changes requested):
I've made some code formmating requests, but the code looks good. Please put a description on the PR.

alla3 commented on src/Makefile.am:
```diff
@@ -10,7 +10,7 @@ libknotcloudsdkc_include_HEADERS = $(lib_headers)
 libknotcloudsdkc_la_SOURCES = $(lib_headers) $(lib_sources)
 libknotcloudsdkc_la_LIBADD = $(modules_libadd) -lm
-libknotcloudsdkc_la_CFLAGS = $(AM_CFLAGS) $(modules_cflags)
+libknotcloudsdkc_la_CFLAGS = $(AM_CFLAGS) $(modules_cflags) -Werror
```
This flag is declared on a 'm4' file, maybe a better approach could be use those macros.

alla3 replied:
Also, it doesn't seem to be related to the commit 17f8583233b9d2044906369f61eaa6a09e3f11b8. I think I would create a new commit with this change.

alla3 commented on src/knot_cloud.c:
```diff
@@ -68,33 +68,12 @@ static void knot_cloud_msg_destroy(struct knot_cloud_msg *msg)
-
-static void *knot_cloud_device_array_foreach(json_object *array_item)
+// criar esse for each la no parser.. passar os params pela chamada
+// schema, id e name.
```
Does this comment should be here?

alla3 commented on src/knot_cloud.c:
```diff
@@ -117,110 +96,77 @@ static int map_routing_key_to_msg_type(const char *routing_key)
+		l_error("Ill-formed JSON message");
+		knot_cloud_msg_destroy(msg);
+		return NULL;
+	}
```
Please, add a blank line before the return statement.

agms-cesar replied:
done

alla3 commented on src/parser.c:
```diff
@@ -95,6 +95,54 @@ static int knot_value_as_int(const knot_value_type *data)
+}
+
+static void *device_array_item(json_object *array_item,
+		  	       create_device_item_cb cb)
```
Please, remove the white spaces between tabs.
```suggestion
			       create_device_item_cb cb)
```

alla3 commented on src/parser.c:
```diff
@@ -95,6 +95,54 @@ static int knot_value_as_int(const knot_value_type *data)
+
+	if (json_object_get_type(jobjkey) != json_type_string)
+		return NULL;
+	
```
Please, remove this boundary tab.
```suggestion

```

alla3 commented on src/parser.c:
```diff
@@ -95,6 +95,54 @@ static int knot_value_as_int(const knot_value_type *data)
+
+	/* Getting 'schema': Mandatory field for registered device */
+	/* FIXME: Call a parser function instead of json-c function*/
+	if (!json_object_object_get_ex(array_item, 
```
Please, remove this white space.
```suggestion
	if (!json_object_object_get_ex(array_item,
```

alla3 commented on src/parser.c:
```diff
@@ -280,78 +354,89 @@ struct l_queue *parser_update_to_list(json_object *jso)
+		l_queue_destroy(list, l_free);
+		return NULL;
+	}
+	
```
Please, remove this tab.
```suggestion

```

alla3 commented on src/parser.c:
```diff
@@ -494,53 +584,68 @@ struct l_queue *parser_queue_from_json_array(json_object *jobj,
+		return NULL;
+	}
 
+	has_err = false;			
```
Please, remove these tabs.
```suggestion
	has_err = false;
```

alla3 commented on src/parser.c:
```diff
@@ -494,53 +584,68 @@ struct l_queue *parser_queue_from_json_array(json_object *jobj,
+			break;
+		}
+	}
+	
```
Please, remove this tab.
```suggestion

```

alla3 commented on src/parser.c:
```diff
@@ -582,74 +689,101 @@ json_object *parser_device_json_create(const char *device_id,
-
-	if (json_object_get_type(jobjkey) != json_type_string)
+	}
+	
```
Please, remove this tab.
```suggestion

```

alla3 commented on src/parser.c:
```diff
@@ -582,74 +689,101 @@ json_object *parser_device_json_create(const char *device_id,
+	if (json_object_get_type(jobjkey) != json_type_string) {
+		return NULL;
+	}
+	
```
Please, remove this tab.
```suggestion

```

alla3 commented on src/parser.c:
```diff
@@ -582,74 +689,101 @@ json_object *parser_device_json_create(const char *device_id,
+	
+	str_key = l_strdup(json_object_get_string(jobjkey));
+	json_object_put(jso);
+	return str_key;
```
Please, add a blank line before return statement.

alla3 commented on src/knot_cloud.c:
```diff
@@ -120,127 +96,77 @@ static int map_routing_key_to_msg_type(const char *routing_key)
+			msg->error = parser_get_key_str_from_json_str(json_str,
+				KNOT_JSON_FIELD_ERROR);
 		}
+	}
```
I like we avoid code repetition but I think this way we should be more carefully in case of an API change. For example, if it is necessary to add a new message type, we must verify if it can impact or not these verification, instead of focus only on the new message.

vitbaq replied:
My only concern is about the change in this function. I understand the reason, but I don't know if the trade-off is interesting.

alla3 commented on src/parser.c:
```diff
@@ -235,45 +286,66 @@ struct l_queue *parser_update_to_list(json_object *jso)
 	int jtype;
 	int olen;
 	uint8_t sensor_id;
+	bool err;
```
I prefer the variable name has_err which was used in others functions such as parser_data_create_object function.

agms-cesar replied:
i agree

alla3 commented on src/parser.c:
```diff
@@ -348,12 +428,15 @@ json_object *parser_data_create_object(const char *device_id, uint8_t sensor_id,
 	json_object_object_add(json_msg, KNOT_JSON_FIELD_DATA, json_array);
 
+	json_str = l_strdup(json_object_to_json_string(json_msg));
+	json_object_put(json_msg);
```
Isn't it necessary to call `json_object_put` for `json_array` and `json_data`?

agms-cesar replied:
No, it isn't. The ownership of those: json_array, json_data belongs to json_msg after json_object_array_add and json_object_object_add

alla3 commented on src/parser.c:
```diff
@@ -400,62 +478,63 @@ struct l_queue *parser_schema_to_list(const char *json_str)
 
 	for (i = 0; i < json_object_array_length(jobjarray); i++) {
-
+		has_err = true;
```
Why the variable has_err is not being used in any verification?

agms-cesar replied:
This variable is not used indeed. I will remove it

agms-cesar commented on src/knot_cloud.c:
```diff
@@ -117,110 +96,77 @@ static int map_routing_key_to_msg_type(const char *routing_key)
+		has_err = msg->device_id ? false : true;
+	}
+
+	if (msg->type == UPDATE_MSG || msg->type == REQUEST_MSG) {
```
this if can be replaced by validating if KNOT_JSON_FIELD_ERROR exists in json_str. If does, validates if field is str or null, otherwise sets error=NULL

agms-cesar commented on src/knot_cloud.c:
```diff
@@ -568,8 +473,7 @@ int knot_cloud_update_schema(const char *id, struct l_queue *schema_list)
-	json_object_put(jobj_schema);
-
+	l_free(json_str);
 	return result;
```
missing blank line before return statemtn

agms-cesar replied:
done

agms-cesar committed:
Removes dependencies from json-c from knot_cloud.

- Changes knot_cloud.c file to not use directly json-c API;
- Modifies parser.c and parser.h methods to wrapper json-c;
- Modifies parser.h callback function to not use json-c;
- Modifies knot_cloud.c device callback function to only create device;

CaioCordeiro commented on src/parser.c:
```diff
@@ -550,29 +657,31 @@ json_object *parser_sensorid_to_json(const char *key, struct l_queue *list)
-
-	return setdatajobj;
+	json_str = l_strdup(json_object_to_json_string(setdatajobj));
+	json_object_put(setdatajobj);
+	return json_str;
```
Its nice to have a blank line before a return statement.

CaioCordeiro commented on src/parser.h:
```diff
@@ -14,24 +14,41 @@
 
-typedef void *(*parser_json_array_item_cb) (json_object *array_item);
+/* JSON Fields (keys) related to KNOT messages */
+#define	KNOT_JSON_FIELD_DEVICE_NAME	"name"
```
Is this tab really necessary?

CaioCordeiro reviewed (changes requested):
LGTM. I've requested only some small code format changes, but the code looks good.

alla3 commented on src/knot_cloud.c:
```diff
@@ -120,127 +96,74 @@ static int map_routing_key_to_msg_type(const char *routing_key)
+
+	if (has_err) {
+		l_error("Ill-formed JSON message");
+		knot_cloud_msg_destroy(msg);
```
Free the allocated memory for the string on `msg->token` inside this function, since the `parser_get_key_str_from_json_str` function is allocating a string.

agms-cesar replied:
in order to do this a had to change token (const char * to char *)

alla3 commented on src/parser.c:
```diff
@@ -365,12 +446,7 @@ json_object *parser_data_create_object(const char *device_id, uint8_t sensor_id,
-	json_object_put(json_array);
-	json_object_put(json_msg);
-	return NULL;
+	return (json_str && !has_err) ? json_str : NULL;
```
The "else" condition should free the memory allocated for `json_string` in case of error.

agms-cesar replied:
done

alla3 commented on src/parser.c:
```diff
@@ -348,12 +426,15 @@ json_object *parser_data_create_object(const char *device_id, uint8_t sensor_id,
 	json_object_object_add(json_msg, KNOT_JSON_FIELD_DATA, json_array);
 
+	json_str = l_strdup(json_object_to_json_string(json_msg));
+	json_object_put(json_msg);
```
I do not think this statement should be executed if the `has_err` flag is set as true.

agms-cesar replied:
done
